File: libs/js/secure.js
// IWX Secure helpers using Web Crypto API (AES-256-GCM)
import { iwxEncodePlain, iwxDecode } from './encoder.js';
import { ERROR_CODES, IwxError } from './errors.js';

const subtle = globalThis.crypto?.subtle || globalThis.crypto?.webcrypto?.subtle;

function toBase64(bytes) {
  if (typeof btoa === 'function') {
    return btoa(String.fromCharCode(...bytes));
  }
  return Buffer.from(bytes).toString('base64');
}

function fromBase64(b64) {
  if (typeof atob === 'function') {
    return Uint8Array.from(atob(b64), (c) => c.charCodeAt(0));
  }
  return Uint8Array.from(Buffer.from(b64, 'base64'));
}

async function importKeyFromBase64(keyB64) {
  if (!subtle) throw new IwxError(ERROR_CODES.SCHEMA_INVALID, 'WebCrypto subtle API unavailable');
  const raw = fromBase64(keyB64);
  return subtle.importKey('raw', raw, { name: 'AES-GCM' }, false, ['encrypt', 'decrypt']);
}

export async function iwxSecureEncryptValues(values, keyB64) {
  const key = await importKeyFromBase64(keyB64);
  const iv = globalThis.crypto.getRandomValues(new Uint8Array(12));
  const plaintext = new TextEncoder().encode(JSON.stringify(values));
  const ciphertext = await subtle.encrypt({ name: 'AES-GCM', iv }, key, plaintext);
  return { iv: toBase64(iv), data: toBase64(new Uint8Array(ciphertext)) };
}

export async function iwxSecureDecryptValues(ivB64, dataB64, keyB64) {
  try {
    const key = await importKeyFromBase64(keyB64);
    const iv = fromBase64(ivB64);
    const raw = fromBase64(dataB64);
    const plaintext = await subtle.decrypt({ name: 'AES-GCM', iv }, key, raw);
    return JSON.parse(new TextDecoder().decode(plaintext));
  } catch (err) {
    throw new IwxError(ERROR_CODES.DECRYPTION_FAILED, 'Unable to decrypt IWX payload', { cause: err });
  }
}

export async function iwxSecureEncode(schemaId, sourceObject, { keyId, keyBase64, algorithm = 'AES-256-GCM', source = 'frontend' } = {}) {
  const plain = iwxEncodePlain(schemaId, sourceObject, { source });
  const encrypted = await iwxSecureEncryptValues(plain._v, keyBase64);
  return { _s: schemaId, _k: keyId, _a: algorithm, _iv: encrypted.iv, _v: encrypted.data };
}

export async function iwxSecureDecode(payload, keyBase64, { target = 'frontend' } = {}) {
  const values = await iwxSecureDecryptValues(payload._iv, payload._v, keyBase64);
  return iwxDecode({ _s: payload._s, _v: values }, { target });
}

File: libs/python/secure.py
import json
from base64 import b64encode, b64decode
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes

from .encoder import iwx_encode_plain
from .decoder import iwx_decode_to_backend, iwx_decode_to_front
from .errors import ERROR_CODES, IWXError


def iwx_secure_encrypt_values(values, key_bytes: bytes):
    """Encrypt an array of values using AES-256-GCM."""
    plaintext = json.dumps(values, separators=(',', ':')).encode('utf-8')
    iv = get_random_bytes(12)
    cipher = AES.new(key_bytes, AES.MODE_GCM, nonce=iv)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)
    data = ciphertext + tag
    return {
        'iv': b64encode(iv).decode('ascii'),
        'data': b64encode(data).decode('ascii'),
    }


def iwx_secure_decrypt_values(iv_b64: str, data_b64: str, key_bytes: bytes):
    try:
        raw_iv = b64decode(iv_b64)
        raw_data = b64decode(data_b64)
        ciphertext, tag = raw_data[:-16], raw_data[-16:]
        cipher = AES.new(key_bytes, AES.MODE_GCM, nonce=raw_iv)
        plaintext = cipher.decrypt_and_verify(ciphertext, tag)
        return json.loads(plaintext.decode('utf-8'))
    except Exception as exc:  # noqa: BLE001
        raise IWXError(ERROR_CODES['DECRYPTION_FAILED'], 'Unable to decrypt IWX payload') from exc


def wrap_secure(schema_id: int, obj: dict, key_bytes: bytes, key_id: str, *, source: str = 'frontend', algorithm: str = 'AES-256-GCM') -> dict:
    plain = iwx_encode_plain(schema_id, obj, source=source)
    encrypted = iwx_secure_encrypt_values(plain['_v'], key_bytes)
    return {'_s': schema_id, '_k': key_id, '_a': algorithm, '_iv': encrypted['iv'], '_v': encrypted['data']}


def unwrap_secure(payload: dict, key_bytes: bytes, *, target: str = 'frontend') -> dict:
    values = iwx_secure_decrypt_values(payload['_iv'], payload['_v'], key_bytes)
    if target == 'backend':
        return iwx_decode_to_backend({'_s': payload['_s'], '_v': values})
    return iwx_decode_to_front({'_s': payload['_s'], '_v': values})
